### PR TITLE
fix(highlight): global ns improvements

### LINKED ISSUE #17420: No way to clear foreground color for DiffDelete with `nvim_set_hl`
### Neovim version (nvim -v)

NVIM v0.7.0-dev+1092-g238b944e5

### Vim (not Nvim) behaves the same?

Not relevant (neovim api)

### Operating system/version

Ubuntu 20.04.3

### Terminal name/version

kitty 0.24.1

### $TERM environment variable

xterm-kitty

### Installation

Build from source

### How to reproduce the issue

Build neovim from source and try to change `DiffDelete` highlight group with:

`:lua vim.api.nvim_set_hl(0, "DiffDelete", { fg = nil, bg = "#ff0000" })` or
`:lua vim.api.nvim_set_hl(0, "DiffDelete", { fg = "NONE", bg = "#ff0000" })` or
`:lua vim.api.nvim_set_hl(0, "DiffDelete", { fg = -1, bg = "#ff0000" })`

### Expected behavior

I would expect at least one of those ways would clear the foreground color of `DiffDelete`

### Actual behavior

It remains `Blue` like it is by default.

### Additional context

I also tried looking in the docs for other ways to only clear the foreground color with the api.
Note that: `:hi! DiffDelete guifg=NONE guibg=#ff0000` works correctly and removes the foreground color

### PR DESCRIPTION
**Pull request by lewis6991**
- `nvim_set_hl(0, <name>, {...})` fully redefines `<name>` instead of modifying/setting individual fields. This means `nvim_set_hl(0, <name>, {})` will clear a highlight.
- `ctermfg` and `ctermbg` no longer fallback to `Normal`.
- Color names can now be passed to `ctermfg` and `ctermbg`. E.g. `ctermfg='blue'`.
- Invalid color names (for gui and cterm) now raise an error.

Fixes #17420

### CODE REVIEW

File: src/nvim/syntax.c [hunk 1]
```diff
@@ -6759,7 +6759,16 @@ void set_hl_group(int id, HlAttrs attrs, Dict(highlight) *dict, int link_id)
   };
 
   for (int j = 0; cattrs[j].dest; j++) {
-    if (cattrs[j].val != -1) {
+    if (cattrs[j].val == -1) {
+      // Should we clear?
+      if ((cattrs[j].name.type == kObjectTypeString
```
Comment (bfredl) [inline src/nvim/syntax.c:6764]: This extra check should be unnecessary, just `*cattrs[j].dest = NULL` directly when `cattrs[j].val == -1`

Comment (lewis6991) [reply src/nvim/syntax.c:6764]: Depends what behaviour we want. I've done this so not passing an attr leaves the original value untouched. Only by passing a negative integer or `'NONE'` clears the attr.

Comment (bfredl) [reply src/nvim/syntax.c:6764]: that's inconsistent with the behavior for `ns>0` tho. The idea of the API is that you can just set the value of a highlight group without worrying whatever was there before.

Comment (lewis6991) [reply src/nvim/syntax.c:6764]: Ok fair enough. I'm not too familiar with `ns>0` behaviour.

File: src/nvim/syntax.c [hunk 2]
```diff
@@ -6758,16 +6758,27 @@ void set_hl_group(int id, HlAttrs attrs, Dict(highlight) *dict, int link_id)
...
+
   for (int j = 0; cattrs[j].dest; j++) {
-    if (cattrs[j].val != -1) {
+    if (cattrs[j].val < 0) {
+      XFREE_CLEAR(*cattrs[j].dest);
+      continue;
+    }
+
+    if (cattrs[j].name.type == kObjectTypeString
+        && cattrs[j].name.data.string.size) {
```
Comment (gpanders) [inline src/nvim/syntax.c:6771]: No need to collapse this, it fits within the 100 char limit

**Comment by gpanders** [comment]
Does this only work for "gui" values? What about `ctermfg` and `ctermbg`?

**Comment by lewis6991** [comment]
I don't think they have the same issue since they don't have corresponding name attrs like the gui fields do.

**Comment by bfredl** [comment]
We could still add the corresponding test for cterm colors to be sure (and so it doesn't break later).

**Comment by gpanders** [comment]
They have the issue that it is not possible to clear them using `nvim_set_hl`. If they are unspecified they just use whatever the `Normal` group is.

```vim
:call nvim_set_hl(0, "CursorLine", #{ctermbg: 11}
:hi CursorLine
CursorLine     xxx ctermfg=7 ctermbg=11
```

There's currently no way to clear either `ctermfg` or `ctermbg`.

File: src/nvim/syntax.c (hunk 2 above)
Comment (lewis6991) [reply src/nvim/syntax.c:6771]: Personally I prefer it this way given how long each term is.

As I do most work in splits on a laptop, I usually keep *most* code within 80 chars as a soft limit, with 100 being the hard limit.

If you insist, I don't mind changing it.

**Comment by gpanders** [comment]
I think this solves the cterm thing:

```diff
diff --git a/src/nvim/highlight.c b/src/nvim/highlight.c
index fcd91cdf1..87670ef0b 100644
--- a/src/nvim/highlight.c
+++ b/src/nvim/highlight.c
@@ -905,14 +905,14 @@ HlAttrs dict2hlattrs(Dict(highlight) *dict, bool use_rgb, int *link_id, Error *e
     hlattrs.rgb_fg_color = fg;
     hlattrs.rgb_sp_color = sp;
     hlattrs.cterm_bg_color =
-      ctermbg == -1 ? cterm_normal_bg_color : ctermbg + 1;
+      ctermbg == -1 ? 0 : ctermbg + 1;
     hlattrs.cterm_fg_color =
-      ctermfg == -1 ? cterm_normal_fg_color : ctermfg + 1;
+      ctermfg == -1 ? 0 : ctermfg + 1;
     hlattrs.cterm_ae_attr = cterm_mask;
   } else {
     hlattrs.cterm_ae_attr = cterm_mask;
-    hlattrs.cterm_bg_color = bg == -1 ? cterm_normal_bg_color : bg + 1;
-    hlattrs.cterm_fg_color = fg == -1 ? cterm_normal_fg_color : fg + 1;
+    hlattrs.cterm_bg_color = bg == -1 ? 0 : bg + 1;
+    hlattrs.cterm_fg_color = fg == -1 ? 0 : fg + 1;
   }
 
   return hlattrs;
```

File: src/nvim/syntax.c (hunk 2 above)
Comment (gpanders) [reply src/nvim/syntax.c:6771]: I understand and appreciate the desire to keep a 80 char limit (I actually pushed against the increase to 100 chars to begin with), but I personally find the two line condition harder to read and parse.

That said this isn't a hill I'm going to die on. It is fine with me -- other folks can chime in if they have an opinion.

Comment (lewis6991) [reply src/nvim/syntax.c:6771]: Ok. I'll change it.

FWIW I think the 100 limit is great for the 5-10% of code that needs it.

**Comment by lewis6991** [comment]
Also made `ctermfg=red` work, which it didn't before.

File: src/nvim/syntax.c [hunk 3]
```diff
@@ -8846,9 +8856,26 @@ RgbValue name_to_color(const char *name)
...
+  int i;
+  int off = TOUPPER_ASC(*name);
+  for (i = ARRAY_SIZE(color_names); --i >= 0;) {
+    if (off == color_names[i][0]
+        && STRICMP(name+1, color_names[i]+1) == 0) {
+      break;
+    }
+  }
+  if (i < 0) {
+    semsg(_("E421: Color name or number not recognized: %s"), name);
```
Comment (gpanders) [inline src/nvim/syntax.c:8874]: This should `return i`, otherwise a negative index is passed to `lookup_color`.

**Comment by lewis6991** [comment]
Also made `ctermfg=redd` and `fg=bleu` raise an error.

**Review by gpanders — APPROVED** [review]

**Comment by echasnovski** [comment]
I wanted to create two issues about `nvim_set_hl()` and global namespace, but found this. Issues doesn't seem to be that big, so decided to put them here (sorry, if it is inconvenient; will create them if you think that is better):
- Allow empty dictionary as `cterm`. This is helpful for automated highlighting creation. Currently this throws a `'cterm' must be a Dictionary` error: `lua vim.api.nvim_set_hl(0, 'Comment', {foreground = 'blue', cterm = {}})`.
- Allow `nocombine` key, as it is a part of global highlights. Currently this throws a `invalid key: nocombine` error: `lua vim.api.nvim_set_hl(0, 'Comment', {foreground = 'blue', nocombine = true})`

**Comment by clason** [comment]
> I wanted to create two issues about `nvim_set_hl()` and global namespace, but found this. Issues doesn't seem to be that big, so decided to put them here (sorry, if it is inconvenient; will create them if you think that is better):

Yes, please open this as a separate issue.

> * Allow `nocombine` key, as it is a part of global highlights. Currently this throws a `invalid key: nocombine` error: `lua vim.api.nvim_set_hl(0, 'Comment', {foreground = 'blue', nocombine = true})`

That is already possible (since https://github.com/neovim/neovim/pull/17377).

**Comment by echasnovski** [comment]
> Yes, please open this as a separate issue.
Ok, sorry.
 
> > * Allow `nocombine` key, as it is a part of global highlights. Currently this throws a `invalid key: nocombine` error: `lua vim.api.nvim_set_hl(0, 'Comment', {foreground = 'blue', nocombine = true})`
> 
> That is already possible (since #17377).

Oh, I thought I tested it with latest nightly, but messed up. Sorry again.

**Pull request merged**